freqtrade/optimize/backtesting.py

```diff
@@ -1,231 +1,302 @@
-# pragma pylint: disable=missing-docstring,W0212
+# pragma pylint: disable=missing-docstring, W0212, too-many-arguments
 
-import logging
-from typing import Dict, Tuple
+"""
+This module contains the backtesting logic
+"""
 
+from typing import Dict, Tuple, Any
 import arrow
 from pandas import DataFrame, Series
 from tabulate import tabulate
 
-import freqtrade.misc as misc
 import freqtrade.optimize as optimize
-from freqtrade import exchange
-from freqtrade.analyze import populate_buy_trend, populate_sell_trend
+from freqtrade.arguments import Arguments
 from freqtrade.exchange import Bittrex
-from freqtrade.main import should_sell
+from freqtrade.configuration import Configuration
+from freqtrade import exchange
+from freqtrade.analyze import Analyze
+from freqtrade.logger import Logger
+from freqtrade.misc import file_dump_json
 from freqtrade.persistence import Trade
-from freqtrade.strategy.strategy import Strategy
-
-logger = logging.getLogger(__name__)
 
 
-def get_timeframe(data: Dict[str, DataFrame]) -> Tuple[arrow.Arrow, arrow.Arrow]:
+class Backtesting(object):
     """
-    Get the maximum timeframe for the given backtest data
-    :param data: dictionary with preprocessed backtesting data
-    :return: tuple containing min_date, max_date
+    Backtesting class, this class contains all the logic to run a backtest
+
+    To run a backtest:
+    backtesting = Backtesting(config)
+    backtesting.start()
     """
-    all_dates = Series([])
-    for pair_data in data.values():
-        all_dates = all_dates.append(pair_data['date'])
-    all_dates.sort_values(inplace=True)
-    return arrow.get(all_dates.iloc[0]), arrow.get(all_dates.iloc[-1])
+    def __init__(self, config: Dict[str, Any]) -> None:
 
+        # Init the logger
+        self.logging = Logger(name=__name__, level=config['loglevel'])
+        self.logger = self.logging.get_logger()
+        self.config = config
+        self.analyze = None
+        self.ticker_interval = None
+        self.tickerdata_to_dataframe = None
+        self.populate_buy_trend = None
+        self.populate_sell_trend = None
+        self._init()
 
-def generate_text_table(
-        data: Dict[str, Dict], results: DataFrame, stake_currency) -> str:
-    """
-    Generates and returns a text table for the given backtest data and the results dataframe
-    :return: pretty printed table with tabulate as str
-    """
-    floatfmt = ('s', 'd', '.2f', '.8f', '.1f')
-    tabular_data = []
-    headers = ['pair', 'buy count', 'avg profit %',
-               'total profit ' + stake_currency, 'avg duration', 'profit', 'loss']
-    for pair in data:
-        result = results[results.currency == pair]
+    def _init(self) -> None:
```

Comment:
I think we can drop `_init`, and initialize `self.analyze` and `exchange._API` directly in the constructor. The Method rebindings aren't really necessary IMO, because it hurts readability.